httpbp/client_middlewares.go

```diff
@@ -0,0 +1,222 @@
+//
+// Adding this middleware means that both, non-nil response and non-nil error,
+// can be returned. If configured the caller needs to call
+// httpbp.DrainAndClose(resp.Body) to ensure the underlying TCP connection can
+// be re-used.
+func ClientErrorWrapper() ClientMiddleware {
+	const limit = 1024
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			resp, err = next.RoundTrip(req)
+			if err != nil {
+				return resp, err
+			}
+			if err == nil {
+				err = ClientErrorFromResponse(resp)
+				if err != nil {
+					defer DrainAndClose(resp.Body)
+					var ce *ClientError
+					if !errors.As(err, &ce) {
+						return resp, err
+					}
+					body, e := io.ReadAll(io.LimitReader(resp.Body, limit))
+					if e != nil {
+						return resp, e
+					}
+					ce.AdditionalInfo = string(body)
+					return resp, ce
+				}
+			}
+			return resp, err
+		})
+	}
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config breakerbp.Config) ClientMiddleware {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return breakerbp.NewFailureRatioBreaker(config)
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+			host := req.URL.Hostname()
+
+			// new circuit breakers should rarely get allocated: the pool reduces
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				defer pool.Put(newBreaker)
+			}
+
+			resp, err := breaker.(breakerbp.FailureRatioBreaker).Execute(func() (interface{}, error) {
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					// read & close to ensure underlying RoundTripper
+					// (http.Transport) is able to re-use the persistent TCP
+					// connection.
+					_, _ = io.ReadAll(resp.Body)
+					return nil, resp.Body.Close()
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+// Retries provides a retry middleware.
+func Retries(retryOptions ...retry.Option) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			if len(retryOptions) == 0 {
+				retryOptions = []retry.Option{retry.Attempts(1)}
+			}
+
+			err = retrybp.Do(req.Context(), func() error {
+				// include ClientErrorWrapper to ensure retry is applied for
+				// some HTTP 5xx responses
+				resp, err = ClientErrorWrapper()(next).RoundTrip(req)
```

Comment:
since we added both `ClientErrorWrapper` and `Retries` into default middlewares list, we don't want to do this here? (do `ClientErrorWrapper` twice is likely harmless, but also wasteful)

Reply:
It is an annoyance to have it run twice but wouldn't it also mean the Retry middleware isn't really useful outside of this scope? I would argue it would be good to leave it in case a user wants to make use of just the retry middleware for their HTTP client.

Reply:
that's fine to me, but if we go that route:

1. We should document both in `Retries` and `ClientErrorWrapper` that `Retries` included `ClientErrorWrapper`.
2. We should remove `ClientErrorWrapper` from default middlewares.
3. The limit size arg we are gonna add to `ClientErrorWrapper` need to be added to `Retries` as well.

Reply:
Ah true, definitely didn't realize we can remove `ClientErrorWrapper` from the default middleware in this case.